src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DCTree.java

```diff
@@ -53,41 +84,150 @@
+     * <li>provided explicitly, such as for subtypes of {@link DCEndPosTree}
+     * </ul>
+     *
+     * The value is relative to the beginning of the comment text.
+     * Use {@link DCDocComment#getSourcePosition(int)} to convert
+     * it to a position in the source file.
+     */
+    public int getEndPosition() {
+        if (this instanceof DCEndPosTree<?> dcEndPosTree) {
+            int endPos = dcEndPosTree.getEndPos();
+
+            if (endPos != NOPOS) {
+                return endPos;
+            }
+        }
+
+        switch (getKind()) {
+            case TEXT -> {
+                DCText text = (DCText) this;
+                return text.pos + text.text.length();
+            }
+
+            case ERRONEOUS -> {
+                DCErroneous err = (DCErroneous) this;
+                return err.pos + err.body.length();
+            }
+
+            case IDENTIFIER -> {
+                DCIdentifier ident = (DCIdentifier) this;
+                return ident.pos + ident.name.length();
+            }
+
+            case AUTHOR, DEPRECATED, HIDDEN, PARAM, PROVIDES, RETURN, SEE, SERIAL, SERIAL_DATA, SERIAL_FIELD, SINCE,
+                    THROWS, UNKNOWN_BLOCK_TAG, USES, VERSION -> {
+                DCTree last = getLastChild();
+
+                if (last != null) {
+                    int correction = (this instanceof DCParam p && p.isTypeParameter && p.getDescription().isEmpty()) ? 1 : 0;
+                    return last.getEndPosition() + correction;
+                }
+
+                String name = ((BlockTagTree) this).getTagName();
+                return this.pos + name.length() + 1;
+            }
+
+            case ENTITY -> {
+                DCEntity endEl = (DCEntity) this;
+                return endEl.pos + endEl.name.length() + 2;
+            }
+
+            case COMMENT -> {
+                DCComment endEl = (DCComment) this;
+                return endEl.pos + endEl.body.length();
+            }
+
+            case ATTRIBUTE -> {
+                DCAttribute attr = (DCAttribute) this;
+                if (attr.vkind == AttributeTree.ValueKind.EMPTY) {
+                    return attr.pos + attr.name.length();
+                }
+                DCTree last = getLastChild();
+                if (last != null) {
+                    return last.getEndPosition() + (attr.vkind == AttributeTree.ValueKind.UNQUOTED ? 0 : 1);
+                }
+            }
+
+            case DOC_COMMENT ->  {
+                DCDocComment dc = (DCDocComment) this;
+                DCTree last = getLastChild();
+                return last == null ? dc.pos : last.getEndPosition();
+            }
+
+            default -> {
+                DCTree last = getLastChild();
+                if (last != null) {
+                    return last.getEndPosition();
+                }
+            }
+        }
+
+        return Position.NOPOS;
```

Comment:
This method should return Diagnostic.NOPOS.

I'm concerned with mixing com.sun.tools.javac.util.Position.NOPOS and javax.tools.Diagnostic.NOPOS. Although they currently have the same int value, they are of different types, used in different contexts, and not guaranteed to remain equal in the future. We might need to do something about it.

Reply:
Hmmm. I doubt this is the only place where the values are conflated.  I'll check with other `javac` folk about defining `com.sun.tools.javac.util.Position.NOPOS` in terms of `javax.tools.Diagnostic.NOPOS`.   The internal version predates the public value, but the two are intended too be the same.

Reply:
I changed the line, but only to remove the redundant use of `Position.`  

The two values, `Diagnostic.NOPOS` and `Position.NOPOS` are intended to be the same value, allowing for the difference in type. The difference in type for `NOPOS`, and all methods that return positions, was intended to allow other implementations of `javax.tools` to handle longer source files (!!!) even though `javac` only support `-int`-sized files.